Instant Search: search from any page on the site

gibrown opened the pull request:
Allow the end user to do instant search and filtering no matter where they are on the site. Filters are always displayed and the end user can always combine filters and search queries.

#### Testing instructions:
* Add define( "JETPACK_SEARCH_PROTOTYPE", true ); to your wp-config.php.
* Ensure that your site has the Jetpack Pro plan and Jetpack Search enabled.
* Add a Jetpack Search widget to the Search page sidebar and configure some filters.
* Here are the cases I've been testing:

```
 - on search page e.g. http://gibrown.wpsandbox.me/main/?blog_id=20115252&orderby=relevance&order=DESC&s=related%20psts
	 - filter
	 - filter then cancel search
	 - filter then search
	 - filter then filter
 - on home page e.g. http://gibrown.wpsandbox.me/main/?blog_id=20115252
	 - filter
	 - search
	 - filter then unfilter
	 - search then esc
	 - filter then search
	 - filter then filter
	 - search then filter
 - on a post page e.g.  http://gibrown.wpsandbox.me/main/?blog_id=20115252&p=1241
	 - filter
	 - filter then unfilter
	 - search
	 - search then esc
	 - filter then search
	 - filter then filter
	 - search then filter

```


#### Proposed changelog entry for your changes:

None.

gibrown committed:
Inserting search results into multiple themes working

gibrown committed:
Get search boxes working in 2013 and improve css and styling.

gibrown committed:
Rename search-widget to search-app

gibrown committed:
Get search box working with twentyten

gibrown committed:
Give up on getting the jp search icon to display. very tricky.

gibrown committed:
Merge remote-tracking branch 'origin/instant-search-master' into add/instant-search-theme-support

gibrown committed:
Fixes based on PR review.

gibrown committed:
Initial implementation of search from anywhere. Not fully working yet.

gibrown committed:
Only debounce the query when the user is changing the search query. No delay for sorting or filtering.

gibrown committed:
Correct initially displaying results.

gibrown committed:
Merge remote-tracking branch 'origin/instant-search-master' into add/instant-search-always-on

gibrown committed:
WIP: fixing bugs

gibrown committed:
Mostly working now

gibrown committed:
Fix search anywhere bugs

gibrown commented on modules/search/instant-search/components/search-app.jsx:
```diff
@@ -35,23 +36,21 @@ class SearchApp extends Component {
 			loading: false,
 		};
-		this.getResults = debounce( this.getResults, 200 );
+		this.getDebouncedResults = debounce( this.getResults, 200 );
```
Note: this is so we can make the UI faster when the user filters/sorts and don't add the extra delay for those cases.

gibrown committed:
Remove onKeyPress

jsnmoon commented:
Discovered a bug:
1. Activate the 2016 theme.
2. Navigate to the home page `/` with the search widget enabled in the sidebar.
3. Enter a query (e.g. `privacy`).

What I expected:
- Search results to render in place of the main content.

What I experienced:
- The main content is erased, but not replaced with search results.

<img width="1247" alt="Screen Shot 2019-10-10 at 1 24 38 PM" src="https://user-images.githubusercontent.com/4044428/66599799-5449f980-eb61-11e9-873a-14aadcf169ba.png">

jsnmoon commented on modules/search/instant-search/components/search-app.jsx:
```diff
@@ -61,56 +60,91 @@ class SearchApp extends Component {
+
+	activateResults() {
+		if ( ! this.state.resultsActive ) {
+			this.setState( { resultsActive: true } );
```
[`setState` is not synchronous](https://reactjs.org/docs/react-component.html#setstate), so we'll want to do something like the following:
```js
this.setState( { resultsActive: true }, () => { doSomethingHere(); } );
```

We might also want `activateResults` to accept a `callback` input like so:
```js
activateResults( callback = () => {} ) {
```

jsnmoon commented on modules/search/instant-search/components/search-app.jsx:
```diff
@@ -61,56 +60,91 @@ class SearchApp extends Component {
+			return;
+		}
+		if ( this.state.resultsActive ) {
+			this.setState( { resultsActive: false } );
```
Ditto here about using [a callback for `setState`](https://reactjs.org/docs/react-component.html#setstate):
`this.setState( { resultsActive: false }, () => { doSomethingHere(); } );`

jsnmoon commented on modules/search/instant-search/components/search-app.jsx:
```diff
@@ -61,56 +60,91 @@ class SearchApp extends Component {
+		if ( this.hasActiveQuery() ) {
+			this.activateResults();
+		}
+		this.maybeDeactivateResults();
```
Shouldn't this line be in an `else` clause?

gibrown replied:
It doesn't actually matter because of then if clause in `maybeDeactivateResults()` but I'll add one to clarify the code.

jsnmoon commented on modules/search/instant-search/components/search-app.jsx:
```diff
@@ -61,56 +60,91 @@ class SearchApp extends Component {
+		if ( this.hasActiveQuery() ) {
+			this.activateResults();
+		}
+		this.maybeDeactivateResults();
```
I'd like to note that both `activateResults` and `maybeDeactivateResults` contain async behavior due to the use of `setState`. In other words, there's a possible race condition here.

gibrown replied:
I think this is better now.

jsnmoon commented on modules/search/instant-search/components/search-results.jsx:
```diff
@@ -45,13 +47,16 @@ class SearchResults extends Component {
+							  )
+							: sprintf( _n( '%s result for "%s"', '%s results for "%s"', total ), num, query )
+						: //only filtering, no search query
+						  sprintf( _n( '%s result', '%s results', total ), num ) }
```
Could we rewrite this to avoid nested ternaries?

jsnmoon reviewed (dismissed):
Two blockers:
- As noted above, there's a bug that prevents search results from loading when searching from the homepage.
- There's a possible race condition in our state handling code, as noted inline below.

gibrown commented on modules/search/instant-search/components/search-app.jsx:
```diff
@@ -35,23 +36,21 @@ class SearchApp extends Component {
 		};
-		this.getResults = debounce( this.getResults, 200 );
+		this.getDebouncedResults = debounce( this.getResults, 200 );
 		this.getResults( this.state.query, getFilterQuery(), this.state.sort );
```
That is too bad you have to wait for the component to mount to start the network request. But ya I guess that makes sense.

I was trying to move the pieces that hide and remove items from the DOM to the constructor because it seemed to speed up. Also we don't load the app until the DOM is rendered so it should be ok to start removing or changing things. Do you see a problem with that? We aren't actually manipulating the DOM for this component.

jsnmoon replied:
[This comment by a React core dev] explaining why this is the case.

> The specific recommendation to avoid side effects in the constructor [...] is related to the changes we are making to allow rendering to be asynchronous and interruptible (in part to support use cases like this better).

jsnmoon replied:
As for the DOM manipulation, I'm stumped given that we're manipulating parts of the DOM outside of the React component tree. I'm cautiously leaning toward keeping it inside the constructor if we can quantify it actually speeds up the user experience.

gibrown replied:
It is definitely hard to quantify. Thanks for that link. Very helpful.

gibrown commented:
> a bug that prevents search results from loading when searching from the homepage

@jsnmoon I'm unable to reproduce a bug here. It searches and loads just fine for me. Is even immediately filtering (without a search) not working? Any errors?

jsnmoon commented:
> > a bug that prevents search results from loading when searching from the homepage
> 
> @jsnmoon I'm unable to reproduce a bug here. It searches and loads just fine for me. Is even immediately filtering (without a search) not working? Any errors?

@gibrown: No errors. Here's a video screencap:

![2019-10-11 16-12-35 2019-10-11 16_13_49](https://user-images.githubusercontent.com/4044428/66688054-37392780-ec42-11e9-9173-507674271245.gif)

gibrown commented:
Yikes, here is what I have tried to reproduce:
- Firefox instead of Chrome
- Set home page to a static page instead of list of posts
- Added widgets to footer
- Added a menu
- Verified that my theme is up to date

None of that made a difference. Could you save the html of your page as a gist? Maybe looking at it will tell me something that I am missing. Or maybe this is one of the race conditions you were warning me about. ;) I'm going to assume it is that and try to just clean those up.

gibrown committed:
Try to prevent possible race conditions.

gibrown committed:
Merge remote-tracking branch 'origin/instant-search-master' into add/instant-search-always-on

gibrown commented:
I am cautiously hopeful that me moving the hideChildren() call to before setState in activateResults() will fix the bug you were seeing.

jsnmoon commented:
> I am cautiously hopeful that me moving the hideChildren() call to before setState in activateResults() will fix the bug you were seeing.

Yup, that seems to have fixed it!

gibrown committed:
Merge remote-tracking branch 'origin/instant-search-master' into add/instant-search-always-on

gibrown commented:
I think one last check and this is ready to go now right?

jsnmoon reviewed (approved):
Yup, I think this is good to go!
